crates/valence_command/src/reader.rs

```diff
@@ -0,0 +1,274 @@
+        ended
+    }
+
+    pub fn read_escaped_str_filtered(
+        &mut self,
+        mut filter: impl FnMut(char) -> StrFilter,
+    ) -> Option<String> {
+        let mut result = String::new();
+        let mut skip_next = false;
+
+        loop {
+            match self.peek_char() {
+                Some(ch) => {
+                    if ch == '\\' && !skip_next {
+                        skip_next = true;
+                        self.cursor += ch.len_utf8();
+                    } else {
+                        match filter(ch) {
+                            StrFilter::Continue => {
+                                self.cursor += ch.len_utf8();
+                                result.push(ch);
+                            }
+                            StrFilter::EndStrInclude => {
+                                self.cursor += ch.len_utf8();
+                                result.push(ch);
+                                break;
+                            }
+                            StrFilter::EndInclude => {
+                                self.cursor += ch.len_utf8();
+                                break;
+                            }
+                            StrFilter::EndExclude => {
+                                break;
+                            }
+                        }
+                        skip_next = false;
+                    }
+                }
+                None => {
+                    return None;
+                }
+            }
+        }
+
+        Some(result)
+    }
+
+    pub fn read_unquoted_str(&mut self) -> &'a str {
+        self.read_str_filtered(|ch| match ch {
+            '0'..='9' | 'A'..='Z' | 'a'..='z' | '_' | '-' | '.' | '+' => StrFilter::Continue,
+            _ => StrFilter::EndExclude,
+        })
+    }
+
+    pub fn read_ident_str(&mut self) -> &'a str {
+        self.read_str_filtered(|ch| match ch {
+            '0'..='9' | 'A'..='Z' | 'a'..='z' | '_' | '-' | '.' | '+' | ':' => StrFilter::Continue,
+            _ => StrFilter::EndExclude,
+        })
+    }
+
+    pub fn read_started_quoted_str(&mut self) -> Option<String> {
+        self.read_escaped_str_filtered(|ch| match ch {
+            '"' | '\'' => StrFilter::EndInclude,
+            _ => StrFilter::Continue,
+        })
+    }
+
+    pub fn skip_started_quoted_str(&mut self) -> bool {
+        self.skip_escaped_str_filtered(|ch| match ch {
+            '"' | '\'' => StrFilter::EndInclude,
+            _ => StrFilter::Continue,
+        })
+    }
+
+    pub fn read_num_str(&mut self) -> &'a str {
+        self.read_str_filtered(|ch| match ch {
+            '0'..='9' | '+' | '-' | 'e' | '.' => StrFilter::Continue,
+            _ => StrFilter::EndExclude,
+        })
+    }
+
+    pub fn read_num<T: FromStr>(&mut self) -> Result<T, <T as FromStr>::Err> {
+        self.read_num_str().parse()
+    }
+
+    pub fn remaining_str(&self) -> &'a str {
+        // SAFETY
+        // cursor is always valid
```

Comment:
nit: would prefer safety comments to all be on the same line like TODO, eg. `// SAFETY: <msg>`